proof/refine/AARCH64/Retype_R.thy

```diff
@@ -5385,6 +5389,45 @@ lemma data_page_relation_retype:
    apply (clarsimp simp: image_def)+
   done
 
+lemma init_arch_objects_APIType_map2_VCPU_noop:
+  "init_arch_objects (APIType_map2 (Inr VCPUObject)) dev ptr n m addrs = return ()"
+  apply (simp add: init_arch_objects_def APIType_map2_def)
+  done
+
+lemma reorder_createObjects_dmo_userPages:
+  "(do
+      addrs <- createObjects y n ko sz;
+      _ <- modify (\<lambda>ks. ks\<lparr>gsUserPages := g ks addrs\<rparr>);
+      _ <- when P (mapM_x (\<lambda>addr. doMachineOp (m addr)) addrs);
+      return (f addrs)
+    od) = (do
+      (addrs, faddrs) <- (do
+        addrs <- createObjects y n ko sz;
+        _ <- modify (\<lambda>ks. ks\<lparr>gsUserPages := g ks addrs\<rparr>);
+        return (addrs, f addrs)
+       od);
+      _ <- when P (mapM_x (\<lambda>addr. doMachineOp (m addr)) addrs);
+      return faddrs
+    od)"
+  by (simp add: bind_assoc)
+
+lemma reorder_createObjects_dmo_gsPTTypes:
+  "(do
+      addrs <- createObjects y n ko sz;
+      _ <- modify (\<lambda>ks. ks\<lparr>ksArchState := gsPTTypes_update (g ks addrs) (ksArchState ks)\<rparr>);
+      _ <- mapM_x (\<lambda>addr. doMachineOp (m addr)) addrs;
+      return (f addrs)
+    od) = (do
+      (addrs, faddrs) <- (do
+        addrs <- createObjects y n ko sz;
+        _ <- modify (\<lambda>ks. ks\<lparr>ksArchState := gsPTTypes_update (g ks addrs) (ksArchState ks)\<rparr>);
+        return (addrs, f addrs)
+       od);
+      _ <- mapM_x (\<lambda>addr. doMachineOp (m addr)) addrs;
+      return faddrs
+    od)"
+  by (simp add: bind_assoc)
```

Comment:
not sure how I'd improve this, but these "reorder" lemmas (that don't actually reorder) are very specific, and a lot of the specifics don't matter; would it be worth at least abstracting these a bit? The second one looks a lot like the first one with P=True, etc.

Reply:
They are more assoc/regroup lemmas than reorder lemmas, I should probably rename. I don't see a good way to abstract them, though, the second one talks about gsPTTypes_update, whereas first one is about gsUserPages with a different form of term pattern. I need to spell out the rest so that g, ks, address etc match the right parts of the rest term. We could make it slightly more abstract by replacing the mapM_x and doMachineOp with variables, but that risks those variables being matched wrong in HO unification. They are really just helpers for extremely specific situations.

Reply:
(have renamed them to `regroup_...` now)